packages/utils/src/dsn.ts

```diff
@@ -1,127 +1,114 @@
+ */
+function dsntoString(dsn: Dsn, withPassword: boolean = false): string {
+  const { host, path, pass, port, projectId, protocol, publicKey } = dsn;
+  return (
+    `${protocol}://${publicKey}${withPassword && pass ? `:${pass}` : ''}` +
+    `@${host}${port ? `:${port}` : ''}/${path ? `${path}/` : path}${projectId}`
+  );
+}
 
-    this._validate();
-  }
+function dsnFromString(str: string): Dsn {
+  const match = DSN_REGEX.exec(str);
 
-  /**
-   * Renders the string representation of this Dsn.
-   *
-   * By default, this will render the public representation without the password
-   * component. To get the deprecated private representation, set `withPassword`
-   * to true.
-   *
-   * @param withPassword When set to true, the password will be included.
-   */
-  public toString(withPassword: boolean = false): string {
-    const { host, path, pass, port, projectId, protocol, publicKey } = this;
-    return (
-      `${protocol}://${publicKey}${withPassword && pass ? `:${pass}` : ''}` +
-      `@${host}${port ? `:${port}` : ''}/${path ? `${path}/` : path}${projectId}`
-    );
+  if (!match) {
+    throw new SentryError('Invalid Dsn');
   }
 
-  /** Parses a string into this Dsn. */
-  private _fromString(str: string): void {
-    const match = DSN_REGEX.exec(str);
+  const [protocol, publicKey, pass = '', host, port = '', lastPath] = match.slice(1);
+  let path = '';
+  let projectId = lastPath;
+
+  const split = projectId.split('/');
+  if (split.length > 1) {
+    path = split.slice(0, -1).join('/');
+    projectId = split.pop() as string;
+  }
 
-    if (!match) {
-      throw new SentryError(ERROR_MESSAGE);
+  if (projectId) {
+    const projectMatch = projectId.match(/^\d+/);
+    if (projectMatch) {
+      projectId = projectMatch[0];
     }
+  }
 
-    const [protocol, publicKey, pass = '', host, port = '', lastPath] = match.slice(1);
-    let path = '';
-    let projectId = lastPath;
+  return dsnFromComponents({ host, pass, path, projectId, port, protocol: protocol as DsnProtocol, publicKey });
+}
 
-    const split = projectId.split('/');
-    if (split.length > 1) {
-      path = split.slice(0, -1).join('/');
-      projectId = split.pop() as string;
-    }
+function dsnFromComponents(components: DsnComponents): Dsn {
+  // TODO this is for backwards compatibility, and can be removed in a future version
+  if ('user' in components && !('publicKey' in components)) {
+    components.publicKey = components.user;
+  }
+
+  return {
+    user: components.publicKey || '',
+    protocol: components.protocol,
+    publicKey: components.publicKey || '',
+    pass: components.pass || '',
+    host: components.host,
+    port: components.port || '',
+    path: components.path || '',
+    projectId: components.projectId,
+  };
+}
 
-    if (projectId) {
-      const projectMatch = projectId.match(/^\d+/);
-      if (projectMatch) {
-        projectId = projectMatch[0];
+function validateDsn(dsn: Dsn): boolean {
+  if (isDebugBuild()) {
```

Comment:
I would invert this statement and change the return value to `void`.

```suggestion
  if (!isDebugBuild()) { return; }
```